File: public/scripts/process_photos.mjs
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';
import { removeBackground } from '@imgly/background-removal-node';

const inputDir = path.resolve('./public/fotos');
const outputDir = path.resolve('./public/fotos_procesadas');

// Create output dir if it doesn't exist
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

// Generate the SVG gradient background
const getGradientSVG = (size) => `
<svg width="${size}" height="${size}" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <radialGradient id="grad1" cx="50%" cy="50%" r="50%" fx="50%" fy="50%">
      <stop offset="0%" style="stop-color:#01273F;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#022038;stop-opacity:1" />
    </radialGradient>
  </defs>
  <rect width="100%" height="100%" fill="url(#grad1)" />
</svg>
`;

async function processImages() {
  const files = fs.readdirSync(inputDir).filter(f => f.startsWith('WhatsApp') && !f.endsWith('.webp'));
  console.log(`Found ${files.length} images to process.`);

  for (const file of files) {
    const inputPath = path.join(inputDir, file);
    const outputName = path.basename(file, path.extname(file)) + '.webp';
    const outputPath = path.join(outputDir, outputName);
    
    console.log(`Processing: ${file}`);
    try {
      // 1. Remove background using @imgly/background-removal-node
      // It expects a file path and returns a Blob. We can convert Blob to ArrayBuffer to Buffer.
      const fileUri = 'file:///' + inputPath.replace(/\\/g, '/');
      const blob = await removeBackground(fileUri);
      const arrayBuffer = await blob.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);

      // 2. Crop to 1:1, resize to a standard size (e.g. 512x512) and Composite with gradient
      const targetSize = 512;
      
      const svgBuffer = Buffer.from(getGradientSVG(targetSize));

      await sharp(buffer)
        .resize(targetSize, targetSize, {
          fit: 'cover',
          position: 'attention' // Smart crop focusing on the subject
        })
        .toBuffer()
        .then(async (croppedSubject) => {
          await sharp(svgBuffer)
            .composite([
              { input: croppedSubject, blend: 'over' }
            ])
            .webp({ quality: 90 })
            .toFile(outputPath);
        });

      console.log(`✅ Success: ${outputName}`);
    } catch (err) {
      console.error(`❌ Failed: ${file}`, err);
    }
  }
}

processImages();

File: public/scripts/process_photos.py
import os
from PIL import Image, ImageDraw, ImageFilter
from rembg import remove
import glob
import random

input_dir = "public/fotos"
output_dir = "public/fotos_procesadas"

os.makedirs(output_dir, exist_ok=True)

def generate_gradient_bg(size=(512, 512)):
    bg = Image.new('RGB', size, (2, 32, 56)) # #022038 edge color
    draw = ImageDraw.Draw(bg)
    
    # Create radial gradient
    center_color = (1, 39, 63) # #01273F
    edge_color = (2, 32, 56)   # #022038
    cx, cy = size[0] // 2, size[1] // 2
    max_radius = (cx**2 + cy**2)**0.5

    for y in range(size[1]):
        for x in range(size[0]):
            distance = ((x - cx)**2 + (y - cy)**2)**0.5
            ratio = min(distance / max_radius, 1.0)
            
            # Interpolate color
            r = int(center_color[0] + (edge_color[0] - center_color[0]) * ratio)
            g = int(center_color[1] + (edge_color[1] - center_color[1]) * ratio)
            b = int(center_color[2] + (edge_color[2] - center_color[2]) * ratio)
            
            # Add subtle noise for texture
            noise = random.randint(-2, 2)
            
            bg.putpixel((x, y), (max(0, min(255, r + noise)), 
                                 max(0, min(255, g + noise)), 
                                 max(0, min(255, b + noise))))
                                 
    # Blur slightly to smooth the gradient/noise
    bg = bg.filter(ImageFilter.GaussianBlur(radius=1))
    return bg

# Find all WhatsApp images
images = glob.glob(os.path.join(input_dir, "WhatsApp*"))
images = [img for img in images if not img.endswith('.webp')]

print(f"Found {len(images)} images to process.")

target_size = (512, 512)

for img_path in images:
    filename = os.path.basename(img_path)
    print(f"Processing: {filename}...")
    try:
        # Open image
        input_image = Image.open(img_path)
        
        # Remove background
        subject = remove(input_image)
        
        # Crop subject to 1:1, centered
        # Get bounding box of the non-transparent pixels
        bbox = subject.getbbox()
        if bbox:
            subject = subject.crop(bbox)
        
        # Make subject fit into target size, keeping aspect ratio
        subject.thumbnail(target_size, Image.Resampling.LANCZOS)
        
        # Generate gradient background
        bg = generate_gradient_bg(target_size)
        
        # Paste subject onto background (centered)
        offset_x = (target_size[0] - subject.width) // 2
        
        # Optional: push the subject down slightly so it sits at the bottom if it's a portrait bust, 
        # but centered vertically is usually fine.
        offset_y = (target_size[1] - subject.height) // 2
        # Let's align to bottom if it's a typical portrait with shoulders cut off:
        offset_y = target_size[1] - subject.height
        
        # Wait, if we align to bottom, there might be empty space at the top.
        # Let's just center it.
        offset_y = (target_size[1] - subject.height) // 2
        
        bg.paste(subject, (offset_x, offset_y), subject)
        
        # Save as webp
        output_name = os.path.splitext(filename)[0] + ".webp"
        output_path = os.path.join(output_dir, output_name)
        
        bg.save(output_path, "WEBP", quality=90)
        print(f"Saved: {output_name}")
    except Exception as e:
        print(f"Failed processing {filename}: {e}")
